### PR TITLE
🔧 Refactor project usage and implement budget rollover ✨

### PR DESCRIPTION
**Pull request by davidreneuw**
## Description

Implemented the azure resource manager sdk for costs, budget and storage management. Implemented a check for "rollover" of budgets: whenever a new fiscal year begins, budgets have their total amount reduced by the amount spent during the previous fiscal year

## How Has This Been Tested?

Wrote unit test for each service and the function. (this is not complete yet)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Unit test (new or update to tests)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (change that would cause documentation to be updated)
- [ ] Configuration change (change that would affect the configuration of the application)

## Checklist:

- [x] PR is submitted to the correct branch `develop`.
- [ ] Code follows the code style of this project.
- [ ] Changes are covered by Specflow tests and all new or existing tests are passing.
- [ ] Any new or updated translatable strings have been added to the localization files.
- [ ] Necessary and relevant documentation has been created or updated.
- [ ] There are either no changes to the application's configuration or the necessary changes in the infrastructure repository are included in a separate PR.

### CODE REVIEW

File: Portal/src/Datahub.Functions/Program.cs [hunk 1]
```diff
@@ -61,6 +84,19 @@
...
+            {
+                options.Diagnostics.IsLoggingEnabled=true;
+                var tenantId = config.GetValue<string>("TENANT_ID");
+                var clientId = config.GetValue<string>("FUNC_SP_CLIENT_ID");
+                var clientSecret = config.GetValue<string>("FUNC_SP_CLIENT_SECRET");
+                var creds = new ClientSecretCredential(tenantId, clientId, clientSecret);
+                var client = new ArmClient(creds);
+                return client;
+            });
+        });
```
Comment (yjmrobert) [inline Portal/src/Datahub.Functions/Program.cs]: What's the difference between the two `services.AddAzureClients` calls?

File: Portal/src/Datahub.Functions/ProjectUsageUpdater.cs [hunk 2]
```diff
@@ -11,38 +12,69 @@ namespace Datahub.Functions;
...
+            else
+            {
+                _logger.LogWarning($"Aborted rollover due to large difference between captured spend and captured costs ({relativeDifference:P})");
+            }
+        }
 
         // queue the usage notification message
-        await _mediator.Send(new ProjectUsageNotificationMessage(message.ProjectId), cancellationToken);
+        await _mediator.Send(new ProjectUsageNotificationMessage(message.ProjectAcronym), cancellationToken);
+        return rolledOver;
```
Comment (yjmrobert) [inline Portal/src/Datahub.Functions/ProjectUsageUpdater.cs:77]: What does the `bool` return value do for a queue triggered function? Is that the status?

File: Portal/src/Datahub.Infrastructure/Services/Cost/WorkspaceBudgetManagementService.cs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+                var projectResources = ctx.Project_Resources2.Include(p => p.Project)
+                    .Where(p => p.Project.Project_Acronym_CD == workspaceAcronym).ToList();
+                var blobType = TerraformTemplate.GetTerraformServiceType(TerraformTemplate.AzureStorageBlob);
+                var blobResource = projectResources.First(r => r.ResourceType == blobType);
+                rgName = ParseResourceGroup(blobResource.JsonContent);
+            }
+
+            var budgetId =
+                $"/subscription/{_armClient.GetDefaultSubscription().Id.SubscriptionId}/resourceGroups/{rgName}/providers/Microsoft.Consumption/budgets/{rgName}-budget";
+            return budgetId;
```
Comment (yjmrobert) [inline Portal/src/Datahub.Infrastructure/Services/Cost/WorkspaceBudgetManagementService.cs]: We'll need a TODO here as I'm going to be attaching the subscription id to a workspace (as there will be multiple subscriptions)

File: Portal/src/Datahub.Portal/Startup.cs [hunk 4]
```diff
@@ -142,6 +146,23 @@ public void ConfigureServices(IServiceCollection services)
...
+                options.Retry.Delay = TimeSpan.FromSeconds(2);
+                var tenantId = Configuration.GetSection("AzureAd").GetValue<string>("TenantId");
+                var clientId = Configuration.GetSection("AzureAd").GetValue<string>("ClientId");
+                var clientSecret = Configuration.GetSection("AzureAd").GetValue<string>("ClientSecret");
+                var subscriptionId = Configuration.GetSection("AzureAd").GetValue<string>("SubscriptionId");
+                var creds = new ClientSecretCredential(tenantId, clientId, clientSecret);
+                var client = new ArmClient(creds, subscriptionId, options);
+                return client;
+            });
+        });
```
Comment (yjmrobert) [inline Portal/src/Datahub.Portal/Startup.cs]: Does the portal need the `ArmClient`?

**Review by yjmrobert — COMMENTED** [review]
Looks good, I'll have to take a peek into the code coverage and make sure we're hitting all the branches.

File: Portal/src/Datahub.Functions/ProjectUsageUpdater.cs (hunk 2 above)
Comment (davidreneuw) [reply Portal/src/Datahub.Functions/ProjectUsageUpdater.cs:77]: I don't believe it does anything functional. This was for testing purposes: the specflow tests hit this method directly and use the return value as part of assertions. The return value is just whether or not the function ran a roll-over

File: Portal/src/Datahub.Functions/Program.cs (hunk 1 above)
Comment (davidreneuw) [reply Portal/src/Datahub.Functions/Program.cs]: None, this is a mistake on my side, will remove one of them

File: Portal/src/Datahub.Portal/Startup.cs (hunk 4 above)
Comment (davidreneuw) [reply Portal/src/Datahub.Portal/Startup.cs]: No, I had preemptively added it as I was planning to hook the budget service to the workspace settings page "budget" configuration. Will remove it for now

File: Portal/test/Datahub.Infrastructure.UnitTests/Services/WorkspaceBudgetManagementServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+using static Datahub.Infrastructure.UnitTests.Testing;
+using NSubstitute;
+
+namespace Datahub.Infrastructure.UnitTests.Services
+{
+    [TestFixture]
+    [Ignore("Missing configuration")]
+    public class WorkspaceBudgetManagementServiceTests
+    {
+        private WorkspaceBudgetManagementService _sut;
```
Comment (yjmrobert) [inline Portal/test/Datahub.Infrastructure.UnitTests/Services/WorkspaceBudgetManagementServiceTests.cs:16]: Just noticed this is called `_sut`? What's that for? lol

**Review by yjmrobert — COMMENTED** [review]
What's needed for this to be merged?

File: Portal/test/Datahub.Infrastructure.UnitTests/Services/WorkspaceBudgetManagementServiceTests.cs (hunk 5 above)
Comment (davidreneuw) [reply Portal/test/Datahub.Infrastructure.UnitTests/Services/WorkspaceBudgetManagementServiceTests.cs:16]: I'm not sure what you mean, `sut` = system under test

Comment (davidreneuw) [reply Portal/test/Datahub.Infrastructure.UnitTests/Services/WorkspaceBudgetManagementServiceTests.cs:16]: Also note this is legacy/unused

**Comment by davidreneuw** [comment]
@yjmrobert Nothing is needed for this to be merged. I just hadn't re-requested your review after last changes

**Pull request merged**